crates/apollo-compiler/src/queries/database.rs

```diff
@@ -0,0 +1,656 @@
+// At the moment all fields in this struct are created here, instead individual
+// queries for selection_set, variables, directives etc can be created.
+fn operations(db: &dyn SourceDatabase) -> Operations {
+    let operations = db
+        .definitions()
+        .iter()
+        .filter_map(|definition| match definition {
+            ast::Definition::OperationDefinition(op_def) => {
+                Some(operation_definition(db, op_def.clone()))
+            }
+            _ => None,
+        })
+        .collect();
+    Operations::new(Arc::new(operations))
+}
+
+fn subscriptions(db: &dyn SourceDatabase) -> Operations {
+    let operations = db
+        .operations()
+        .iter()
+        .filter_map(|op| {
+            if op.ty == OperationType::Subscription {
+                Some(op.clone())
+            } else {
+                None
+            }
+        })
+        .collect();
+    Operations::new(Arc::new(operations))
+}
+
+fn mutations(db: &dyn SourceDatabase) -> Operations {
+    let operations = db
+        .operations()
+        .iter()
+        .filter_map(|op| {
+            if op.ty == OperationType::Mutation {
+                Some(op.clone())
+            } else {
+                None
+            }
+        })
+        .collect();
+    Operations::new(Arc::new(operations))
+}
+
+fn query(db: &dyn SourceDatabase) -> Operations {
+    let operations = db
+        .operations()
+        .iter()
+        .filter_map(|op| {
+            if op.ty == OperationType::Query {
+                Some(op.clone())
+            } else {
+                None
+            }
+        })
+        .collect();
+    Operations::new(Arc::new(operations))
+}
+
+fn find_operation(db: &dyn SourceDatabase, id: Uuid) -> Option<Arc<OperationDefinition>> {
+    db.operations().iter().find_map(|op| {
+        if id == op.id() {
+            return Some(Arc::new(op.clone()));
+        }
+        None
+    })
+}
+
+fn operation_definitions_names(db: &dyn SourceDatabase) -> Arc<Vec<String>> {
+    Arc::new(db.operations().iter().filter_map(|n| n.name()).collect())
+}
+
+// NOTE: potentially want to return a hashmap of variables and their types?
+fn operation_definition_defined_variables(
+    db: &dyn SourceDatabase,
+    id: Uuid,
+) -> Option<Arc<HashSet<String>>> {
+    let vars: HashSet<String> = db
+        .find_operation(id)?
+        .variables()?
+        .iter()
+        .map(|v| v.name())
+        .collect();
+    Some(Arc::new(vars))
+}
+
+fn operation_fields(db: &dyn SourceDatabase, id: Uuid) -> Option<Arc<Vec<Field>>> {
+    let fields: Vec<Field> = db
+        .find_operation(id)?
+        .selection_set()
+        .iter()
+        .filter_map(|sel| match sel {
+            Selection::Field(field) => Some(field.as_ref().clone()),
+            _ => None,
+        })
+        .collect();
+    Some(Arc::new(fields))
+}
+
+fn operation_inline_fragment_fields(db: &dyn SourceDatabase, id: Uuid) -> Option<Arc<Vec<Field>>> {
```

Comment:
I wonder if instead of having this kind of long function name we could implement method like `fields(&self, &dyn SourceDatabase) -> Option<Arc<Vec<Field>>>`

BTW why should we put this into an `Arc` ? Maybe we should let the user choose, if (s)he needs to wrap the value to an `Arc` then the user could do it by (her/him)self.

Last thought, why not just returning a `Vec<Field>` instead of an `Option` ? I mean if the Vec is empty it means there is no fields ? I don't know, just some thoughts. Would be happy to discuss with you about that

Reply:
@bnjjj appreciate the effort in trying to review this! I think at this point of this work it's not super useful for me to have the API reviewed, however. It's definitely far from the final form of what users are going to be working with, or what we will be publishing. I wanted to have something rounded up and merged so I am not continuously working on the same branch. If you are keen to review, I can let you know when the crate is ready for a proper look over and the sorts of feedback I am looking for!

> I wonder if instead of having this kind of long function name we could implement method like fields(&self, &dyn SourceDatabase) -> Option<Arc<Vec<Field>>>

this function is not intended to be used directly, but is meant to be memoised in the db for various validation rules to use. and yep, there is a plan to have a .fields() function for `InlineFragment` value for better ergonomics

> BTW why should we put this into an Arc ? Maybe we should let the user choose, if they needs to wrap the value to an Arc then the user could do it by themselves.

There is a lot of things here with Arc, but the intent is for the users to never use this directly. These queries are simply here to be memoised for either our own compiler uses or for the wrapper API. This is very likely to change quite a bit next week as I'll be breaking down the database into smaller chunks, and fixing the wrapper API, so it's a bit early to discuss this fully. 

> Last thought, why not just returning a Vec<Field> instead of an Option ? I mean if the Vec is empty it means there is no fields?

Could probably return an empty vec \o/. I'll have to rewrite this function once the databases get split, so this will change also 😬